contracts/lib.rs

```diff
@@ -469,258 +462,257 @@ pub mod prosopo {
+            fee: u32,
+            payee: Payee,
+        ) -> Result<(), Error> {
+            // this function is for updating only, not registering
+            if self.providers.get(self.env().caller()).is_none() {
+                return err!(Error::ProviderDoesNotExist);
             }
 
-            // Get provider
-            let mut provider = self
-                .providers
-                .get(provider_account)
-                .ok_or_else(err_fn!(Error::ProviderDoesNotExist))?;
-
-            // Update provider status
-            self.provider_change_status(
-                provider_account,
-                provider.status,
-                GovernanceStatus::Deactivated,
-                provider.payee,
-            );
-            provider.status = GovernanceStatus::Deactivated;
-            self.providers.insert(provider_account, &provider);
+            self.provider_configure(
+                Some(service_origin),
+                Some(fee),
+                Some(payee),
+                false,
+                None,
+                None,
+            )
+        }
 
-            Ok(())
+        /// De-Register a provider by setting their status to Deactivated
+        #[ink(message)]
+        pub fn provider_deactivate(&mut self) -> Result<(), Error> {
+            // Change status to deactivated
+            self.provider_configure(None, None, None, true, None, None)
         }
 
         /// Unstake and deactivate the provider's service, returning stake
         #[ink(message)]
         #[ink(payable)]
-        pub fn provider_unstake(&mut self) -> Result<(), Error> {
-            let caller = self.env().caller();
-            if self.providers.get(caller).is_none() {
-                return err!(Error::ProviderDoesNotExist);
-            }
+        pub fn provider_deregister(&mut self) -> Result<(), Error> {
+            let provider_account = self.env().caller();
 
-            let provider = self.get_provider_details(caller)?;
+            let provider = self.get_provider(provider_account)?;
+
+            // remove the provider
+            self.providers.remove(provider_account);
+
+            // remove the provider from their category
+            self.provider_state_remove(&provider)?;
+
+            // return the stake
             let balance = provider.balance;
             if balance > 0 {
                 self.env()
-                    .transfer(caller, balance)
+                    .transfer(provider_account, balance)
                     .map_err(|_| Error::ContractTransferFailed)?;
-                self.provider_deregister(caller)?;
             }
 
             Ok(())
         }
 
+        fn get_provider(&self, account: AccountId) -> Result<Provider, Error> {
+            self.providers
+                .get(account)
+                .ok_or_else(err_fn!(Error::ProviderDoesNotExist))
+        }
+
+        #[ink(message)]
+        pub fn provider_fund(&mut self) -> Result<(), Error> {
```

Comment:
`If self.env.transferred > 0`?

Reply:
I can't see exactly where you mean, but I've got a check for this in provider_configure() now to do the diff, so the balance transfer >0 will also be checked there